### ISSUE TITLE
`docker` module image pulling broken in `devel`

### ISSUE THREAD

**Issue by bensymonds** [ansible/ansible#8925]
##### Issue Type:

“Bug Report”
##### Ansible Version:

`ansible 1.8`
`devel`@`1711df3b75dd226bbb2b2592ca3ee5165c311e17`
##### Environment:

N/A
##### Summary:

When you use the `docker` module to run an image, the image should be pulled from registry if necessary. This was the behaviour in v1.7.1, but seems to be broken in `devel`.

I would blame this line: https://github.com/ansible/ansible/compare/v1.7.1...devel#diff-c2983e7efb083bc8a8ce86d6b1bcd059R629. `resource` is the combined image name and tag, whereas it should be the image name (hopefully that makes sense to you - the variables/params in this module are badly named).
##### Steps To Reproduce:

Use `docker` module to run an image which is not present on the machine.
##### Expected Results:

The image should be pulled from registry and then run.
##### Actual Results:

The module fails with an error: `failed to pull the specified image`.

**Comment by iragsdale** [comment]
I'm seeing this issue as well.

**Comment by abadger** [comment]
@bensymonds, @iragsdale What command are you running?  I'm trying to replicate this but what I'm trying is going down a different code path....

**Comment by iragsdale** [comment]
Just running a playbook that includes a task that looks like this:

```
- name: create BCC web instances
  sudo: yes
  with_items: XXX
  register: XXX
  docker:
    name: XXXX
    state: running
    image: XXXXX
    env:
      XXXX: XXXX
    expose: 3000
    command: XXXX
```

If the image doesn't already exist on the box, then it fails completely. If I manually pull the image, it works fine.

**Comment by iragsdale** [comment]
This is from the devel branch, last commit was 8cc354

**Comment by abadger** [comment]
I tried this:

```
ansible localhost -m docker -a 'image=fedora/ssh state=running command=hostname'
```

which worked with current devel.

**Comment by abadger** [comment]
Thanks, testing.

**Comment by bensymonds** [comment]
Taking a very quick look at the code - I think the problem may only occur if you specify a tag in the image name. So try with `image=fedora/ssh:latest` instead of just `image=fedora/ssh`.

**Comment by abadger** [comment]
@bensymonds Thanks, that did it.  Looking for the best way to fix it now.

**Comment by abadger** [comment]
Looks like we're calling self.client.pull(image, tag = tag) with the same values (fedora/ssh, latest in my test) for both 1.7.1 and devel.  So it's something about how we're initializing the docker library that's doing this to us.

**Comment by bensymonds** [comment]
@abadger apologies, I think it works when the tag is `latest` (I hadn't actually tested that case). Try with a different tag, e.g. `busybox:ubuntu-14.04`.

**Comment by abadger** [comment]
This was introduced in the fix for issue #8670 which changed parsing of the image to account for organizations and private registries.  The fix here will update the parsing again so that we'll still include organization and private registry information but omit the tag from the image name.

**Comment by abadger** [comment]
# Closing This Ticket

Hi!

We believe recent commits (likely detailed above) should resolve this question or problem for you.

This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
